pandas/core/categorical.py

```diff
@@ -1665,16 +1669,33 @@ def fillna(self, value=None, method=None, limit=None):
 
         else:
 
-            if not isna(value) and value not in self.categories:
-                raise ValueError("fill value must be in categories")
+            # If value is a dict or a Series (a dict value has already
+            # been converted to a Series)
+            if isinstance(value, ABCSeries):
+                if not value[~value.isin(self.categories)].isna().all():
+                    raise ValueError("fill value must be in categories")
+
+                values_codes = _get_codes_for_values(value, self.categories)
+                indexer = np.where(values_codes != -1)
+                values[indexer] = values_codes[values_codes != -1]
+
+            # If value is not a dict or Series it should be a scalar
+            elif is_scalar(value):
+                if not isna(value) and value not in self.categories:
+                    raise ValueError("fill value must be in categories")
+
+                mask = values == -1
+                if mask.any():
+                    values = values.copy()
+                    if isna(value):
+                        values[mask] = -1
+                    else:
+                        values[mask] = self.categories.get_loc(value)
 
-            mask = values == -1
-            if mask.any():
-                values = values.copy()
-                if isna(value):
-                    values[mask] = -1
-                else:
-                    values[mask] = self.categories.get_loc(value)
+            else:
+                raise TypeError('"value" parameter must be a scalar, dict '
```

Comment:
do we have testing to hit this?

Reply:
The problem is that I can't think of an example where we would actually hit this specific line. If the user passed a list, tuple, or DataFrame as a value in ``fillna`` this will be caught by ``fillna`` in ``generic.py`` rather than in ``categorical.py`` with an identical error message to the one here. So in some ways this exception is kind of redundant because all of the work should be done in ``generic.py``. But it might be useful to keep it anyway.

I have included tests in ``test_fillna_categorical_raise()`` in ``tests/series/test_missing.py`` to check for the ``TypeError`` when the user passes a list, tuple, or dict fill ``value``. But the tests aren't parametrized.